### DISCUSSION TITLE
Performance drop with IPSec tunnel compared to the performance of the link without IPSec tunnel.

### DISCUSSION THREAD

**Discussion by hbhokare** [strongswan/strongswan#870] [General]
We are facing a performance issue with IPSec for MTU size = 1500 with both AES-NI and Crypto offloaded enabled runs.  We are trying to find out configurations that we can do to strongswan, IPSec stack that can improve this performance. We have seen a significant performance drop with IPSec tunnel compared to the performance of the link without IPSec tunnel.
Here are the setup details, swanctl config, iperf results with and without IPsec:

Platform and Hardware config:
Kernel: 4.18.0-305.25.1.el8
CPU: 16 core Intel(R) Xeon(R) Bronze 3106 CPU @ 1.70GHz with AESNI support
MTU: 1500
Ethernet interface: 25Gbps chelsio link
IPsec performance measured with default iperf settings:
server: iperf3 -s -B 192.168.56.24 -p 4500
client: iperf3 -c 192.168.56.24 -t 100   -p 4500  -B 192.168.56.22
swanctl config :
....
               children {
                   net-net {
                       replay_window = 1024
                       start_action = trap
                       esp_proposals = aes128-sha256-esn,aes128gcm16
                       mode = tunnel
                   }
....
Observations:
Node69:~ # ip xfrm state show
src 192.168.56.22 dst 192.168.56.24
        proto esp spi 0xc6b125f9 reqid 1 mode tunnel
        replay-window 0 flag af-unspec esn
        auth-trunc hmac(sha256) 0xac37fde6971d204260213a01667132cef3f0bcc0edb181904294c8f9cbd55b34 128
        enc cbc(aes) 0x378f07a61a5ba20c8e8773e47b604e6c
        anti-replay esn context:
         seq-hi 0x0, seq 0x0, oseq-hi 0x0, oseq 0x3378fd
         replay_window 1, bitmap-length 1
         00000000
src 192.168.56.24 dst 192.168.56.22
        proto esp spi 0xc08af3ac reqid 1 mode tunnel
        replay-window 0 flag af-unspec esn
        auth-trunc hmac(sha256) 0x46b567d50ddb4fe038405e7f9c75691f0769c66494edcad29ddf88e270b6693d 128
        enc cbc(aes) 0x6420abdf9c68962b99f92bfd721aee39
        anti-replay esn context:
         seq-hi 0x0, seq 0x1d101, oseq-hi 0x0, oseq 0x0
         replay_window 128, bitmap-length 4
         ffffffff ffffffff ffffffff ffffffff
similar output on other side of the tunnel.
Results:
1. 20.5 Gbits/sec unencrypted without IPsec
2. 201 Mbits/sec IPsec (aes128-sha256-esn) with replay_window = 0
3. 304 Mbits/sec IPSec (aes128-sha256-esn) with replay_window = 32
4. 304 Mbits/sec IPsec (aes128-sha256-esn) with replay window = 128
5. 310 Mbits/sec IPsec (aes128-sha256-esn) with replay window = 1024
We have tried :
  - configuring higher replay window size: 1000 instead of default32
  - configuring mss to not have fragmentation
  - we are also exploring if cryptd config cryptd_qlen would have any impact
We need help to understand what's going wrong here. Are we not configuring these settings correctly? And are there any other parameters we can tune to improve this performance?

**Comment by tobiasbrunner** [comment]
You don't want to use SHA-256 if performance is your goal. While AES is accelerated by AES-NI, HMAC-SHA256 is not (and even if you'd have hardware acceleration for the latter, it still requires a non-ideal sequential order for the two algorithms). So always make sure to use AES-GCM on such platforms, then you will fully benefit from AES-NI and PCLMULQDQ instructions.

While you should definitely be able to achieve more than 300 Mbps, the kernel's IPsec implementation still has limitations (max. per single SA is in the range of 2-5 Gbps). Also depends on the number of flows, the size of the packets (i.e. you should use a higher MTU than 1500 if possible), if traffic is forwarded or processed locally, offloading features by the NIC, and general networking kernel knobs (e.g. for [RSS/RPS](https://www.kernel.org/doc/Documentation/networking/scaling.txt)). Unfortunately, I personally don't have any experience with this, so I can't really help you getting that right.
